### PR TITLE
PIM-8789: Merge 3.2 into master

### PR DESCRIPTION
**Pull request by rouliane**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/PdfGeneration/Renderer/ProductPdfRenderer.php [hunk 1]
```diff
@@ -80,6 +80,7 @@ public function __construct(
         $this->uploadDirectory = $uploadDirectory;
         $this->attributeOptionRepository = $attributeOptionRepository;
         $this->customFont = $customFont;
+        $this->attributeOptionRepository = $attributeOptionRepository;
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/PdfGeneration/Renderer/ProductPdfRenderer.php]: Add CHANGELOG

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductIdentifiers.php [hunk 2]
```diff
@@ -27,6 +27,8 @@ public function __construct(Connection $connection)
 
     public function fetchByProductIdentifiers(array $productIdentifiers): array
     {
+        // TODO TIP-1231: remove the 'IF' conditions in the JSON_MERGE_PRESERVE statement,
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductIdentifiers.php]: TODO (you can do it on master, it's easy)
(I can help you with this one)

File: src/Akeneo/Pim/Enrichment/Bundle/Resources/config/queries.yml [hunk 3]
```diff
@@ -1,296 +1,301 @@
 parameters:
-    pim_catalog.doctrine.query.find_variant_product_completeness.class: 'Akeneo\Pim\Enrichment\Bundle\Doctrine\ORM\Query\VariantProductRatio'
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Resources/config/queries.yml]: This file is unreadable. Can you put back the indentation ?

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/ProductGrid/FetchProductModelRowsFromCodes.php [hunk 4]
```diff
@@ -248,7 +248,10 @@ private function getValueCollection(array $codes, array $attributeCodes, string
             SELECT 
                 pm.code,
                 a_label.code attribute_as_label_code,
-                JSON_MERGE(COALESCE(parent.raw_values, '{}'), pm.raw_values) as raw_values
+                JSON_MERGE(
+                    CASE parent.raw_values WHEN '[]' THEN '{}' ELSE COALESCE(parent.raw_values, '{}') END,
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/ProductGrid/FetchProductModelRowsFromCodes.php]: Remove the CASE WHEN, just put COALESCE(parent.raw_values, '{}')

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/ProductGrid/FetchProductRowsFromIdentifiers.php [hunk 5]
```diff
@@ -123,7 +123,11 @@ private function getValueCollection(array $identifiers, array $attributeCodes, s
                 p.identifier,
                 a_label.code attribute_as_label_code,
                 a_image.code attribute_as_image_code,
-                JSON_MERGE(COALESCE(pm1.raw_values, '{}'), COALESCE(pm2.raw_values, '{}'), p.raw_values) as raw_values
+                JSON_MERGE(
+                   CASE pm1.raw_values WHEN '[]' THEN '{}' ELSE COALESCE(pm1.raw_values, '{}') END,
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/ProductGrid/FetchProductRowsFromIdentifiers.php]: Same

File: src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/attribute.yml [hunk 6]
```diff
@@ -67,7 +67,7 @@ datagrid:
             columns:
                 code:
                     type:      string
-                    label:     pim_common.code
+                    label:     Code
```
Comment (pierallard) [inline src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/attribute.yml]: put it back

File: src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/family.yml [hunk 7]
```diff
@@ -52,7 +52,7 @@ datagrid:
             columns:
                 code:
                     type: string
-                    label: pim_common.code
+                    label: Code
```
Comment (pierallard) [inline src/Akeneo/Pim/Structure/Bundle/Resources/config/datagrid/family.yml]: same

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/less/components/LoadingMask.less [hunk 8]
```diff
@@ -6,6 +6,27 @@
     100%{background-position:0% 50%}
 }
 
+.AknLoadingPlaceHolderContainer > * {
```
Comment (pierallard) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/public/less/components/LoadingMask.less]: wtf......

File: src/Akeneo/UserManagement/Bundle/Controller/Rest/UserController.php [hunk 9]
```diff
@@ -391,7 +398,7 @@ private function validatePassword(UserInterface $user, $data): ConstraintViolati
             );
         }
         if (
-            isset($data['new_password']) &&  strlen($data['new_password']) < 2
+            isset($data['new_password']) && strlen($data['new_password']) < 2
         ) {
             $violations[] = new ConstraintViolation(
                 $this->translator ?
```
Comment (pierallard) [inline src/Akeneo/UserManagement/Bundle/Controller/Rest/UserController.php]: Remove this condition

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/ElasticsearchAndSql/FollowUp/GetCompletenessPerChannelAndLocale.php [hunk 10]
```diff
@@ -87,7 +86,16 @@ private function getCategoriesCodesAndLocalesByChannel(string $translationLocale
                         child.root
                 ) AS child ON root.id = child.root_id
                 JOIN pim_catalog_channel as channel ON root.id = channel.category_id
-                LEFT JOIN pim_catalog_channel_translation as channel_translation ON channel.id = channel_translation.foreign_key AND channel_translation.locale = :locale
+                LEFT JOIN
+                (
+                    SELECT
+                        channel.code as channel_code,
+                        JSON_OBJECTAGG(channel_translation.locale, channel_translation.label) as json_labels
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/ElasticsearchAndSql/FollowUp/GetCompletenessPerChannelAndLocale.php]: there is a bug here if there is not translation, as you can't aggregated on with a null key:

[ null => null] is not a valid Json and mysql will throw an error

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductController.php [hunk 11]
```diff
@@ -138,6 +141,7 @@ public function __construct(
         NormalizerInterface $constraintViolationNormalizer,
         ProductBuilderInterface $variantProductBuilder,
         AttributeFilterInterface $productAttributeFilter,
+        Client $productClient,
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductController.php]: to remove, it does no exist anymore there is only one index left

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductController.php [hunk 12]
```diff
@@ -328,6 +333,7 @@ public function removeAction(Request $request, $id)
         $product = $this->findProductOr404($id);
         $this->productRemover->remove($product);
 
+        $this->productClient->refreshIndex();
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductController.php]: to remove

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductModelController.php [hunk 13]
```diff
@@ -88,6 +88,9 @@ class ProductModelController
     /** @var AttributeFilterInterface */
     private $productModelAttributeFilter;
 
+    /** @var Client */
+    private $productModelClient;
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductModelController.php]: to remove

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductModelController.php [hunk 14]
```diff
@@ -109,8 +112,8 @@ class ProductModelController
      * @param NormalizerInterface               $violationNormalizer
      * @param FamilyVariantRepositoryInterface  $familyVariantRepository
      * @param AttributeFilterInterface          $productModelAttributeFilter
+     * @param Client                            $productModelClient
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/InternalApi/ProductModelController.php]: to remove (everywhere)

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductIdentifiers.php (hunk 2 above)
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductIdentifiers.php]: not needed anymor, it's clean on master, you can't have this bug anymore

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductIdentifiers.php [hunk 15]
```diff
@@ -37,7 +39,11 @@ public function fetchByProductIdentifiers(array $productIdentifiers): array
     p.updated,
     f.code AS family_code,
     JSON_ARRAYAGG(g.code) AS group_codes,
-    JSON_MERGE(COALESCE(pm1.raw_values, '{}'), COALESCE(pm2.raw_values, '{}'), p.raw_values) as raw_values
+    JSON_MERGE_PRESERVE(
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductIdentifiers.php]: no need to pull-up this line anymore.

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/ProductGrid/FetchProductModelRowsFromCodes.php [hunk 16]
```diff
@@ -248,7 +248,10 @@ private function getValueCollection(array $codes, array $attributeCodes, string
             SELECT 
                 pm.code,
                 a_label.code attribute_as_label_code,
-                JSON_MERGE(COALESCE(parent.raw_values, '{}'), pm.raw_values) as raw_values
+                JSON_MERGE(
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/ProductGrid/FetchProductModelRowsFromCodes.php]: keep master, no need to pull-up it

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/ProductGrid/FetchProductRowsFromIdentifiers.php [hunk 17]
```diff
@@ -123,7 +123,11 @@ private function getValueCollection(array $identifiers, array $attributeCodes, s
                 p.identifier,
                 a_label.code attribute_as_label_code,
                 a_image.code attribute_as_image_code,
-                JSON_MERGE(COALESCE(pm1.raw_values, '{}'), COALESCE(pm2.raw_values, '{}'), p.raw_values) as raw_values
+                JSON_MERGE(
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/ProductGrid/FetchProductRowsFromIdentifiers.php]: same

File: tests/back/Pim/Enrichment/EndToEnd/Product/Product/ExternalApi/PartialUpdateProductEndToEnd.php [hunk 18]
```diff
@@ -171,6 +171,72 @@ public function testProductCreationWithDifferentIdentifiers()
         $this->assertSame(Response::HTTP_UNPROCESSABLE_ENTITY, $response->getStatusCode());
     }
 
+    public function testProductCreationWithInvalidValues()
```
Comment (ahocquard) [inline tests/back/Pim/Enrichment/EndToEnd/Product/Product/ExternalApi/PartialUpdateProductEndToEnd.php]: useless integration test in my opinion...

File: tests/back/Pim/Enrichment/EndToEnd/Product/Product/ExternalApi/PartialUpdateProductEndToEnd.php [hunk 19]
```diff
@@ -171,6 +171,72 @@ public function testProductCreationWithDifferentIdentifiers()
...
+        ];
+
+        $client->request('PATCH', 'api/rest/v1/products/new_identifier', [], [], [], $data);
+
+        $response = $client->getResponse();
+        $this->assertSame($expectedContent, json_decode($response->getContent(), true));
+        $this->assertSame(Response::HTTP_UNPROCESSABLE_ENTITY, $response->getStatusCode());
+    }
+
+    public function testProductUpdateWithInvalidValues()
```
Comment (ahocquard) [inline tests/back/Pim/Enrichment/EndToEnd/Product/Product/ExternalApi/PartialUpdateProductEndToEnd.php]: same, please remove it

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductIdentifiers.php (hunk 2 above)
Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductIdentifiers.php]: just need to keep the thing as it is on master isn't it @pierallard ? As you fixed everything.

File: CHANGELOG-3.2.md [hunk 20]
```diff
@@ -55,7 +73,6 @@
 ## Bug fixes
 
 - PIM-8270: Update export jobs after a change on a channel category
-- GITHUB-10314: Fix family locale on mass edit product - cheers @gauquier!
```
[the remark this answers is not in the document]
Comment (rouliane) [reply CHANGELOG-3.2.md]: It's not here in the 3.2, so there is no mistake. The 3.2 branch is the singe source of truth regarding the 3.2 changelog

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/less/components/LoadingMask.less (hunk 8 above)
Comment (rouliane) [reply src/Akeneo/Platform/Bundle/UIBundle/Resources/public/less/components/LoadingMask.less]: Sorry I didn't write this. I let you check with the team who wrote this code.

**Pull request merged**